client/packages/programs/src/JsonForms/useFormActions.tsx

```diff
@@ -0,0 +1,86 @@
+/**
+ * Some JSONForm components can have actions associated with them, which are
+ * executed when the whole form is saved/submitted. This hook provides a
+ * mechanism to "register" such actions. They can be defined to be run either
+ * *before* or *after* the main "Save" action (in useJsonForms)
+ *
+ * This hook is only called from `useJsonForms` and the "formActions" object it
+ * returns is passed to all form components on the main `config` prop. This
+ * ensures all form components are sharing a comment Actions registry.
```

Comment:
Another thought was to create this as a Zustand store and get components that need it to access it directly via a hook, rather than passing down through JsonForms component.